### PR TITLE
#9396 Review annotations UI and system

### PR DESCRIPTION
**Pull request by allyoucanmap**
### Main changes

The Annotations plugin has been refactored to include the support in 3D Maps, below the major updates introduced:

- The feature collections included inside a single annotations layer per map has been converted to single MapStore layers. This change allow to manage annotations layers among the other layer inside the TOC. This change allowed to remove the additional panel with list of annotations because the annotations will be listed with other layers.
- The UX of annotation edit panel has been changed with the aim to reduce as much as possible confirm modals or disabled state that could block the editing workflow. An history local reducer has been introduced in the annotation feature editor component to undo modification to annotations (see video below)
- Aligned styles of editing feature between 2D/3D view. Unfortunately some modify interaction behave differently in 2D and 3D like dragging vs clicking to modify coordinates nodes
- Centralized as much as possible the annotations code inside a plugin module. There is still some code to remove but it will be manage in a different issue
- Measurement features editing has been improved inside the annotation editor panel. It is possible to recognize measurement annotations because of a different glyph. Every change on the measure annotation will update the measurement values (see video below)
- the `geosedic` and `fields` properties has been marked as **deprecated** because they need to be reviewed separately

### Add workflow

https://github.com/geosolutions-it/MapStore2/assets/19175505/52078182-420a-4054-9df2-17110d57c1d5

### Add measure to annotations workflow

https://github.com/geosolutions-it/MapStore2/assets/19175505/6614eb15-3a8e-4f3f-a38e-1cf40617ba35

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)

 - [x] Feature and Redactor

## Issue

**What is the current behavior?**

#9396

**What is the new behavior?**

Editing of annotations is available also in 3D maps

This PR refactors the annotations to add support in 3D maps

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)

 - [x] Yes

## Other useful information
**Further enhancements will be provided** with https://github.com/geosolutions-it/MapStore2/issues/9410

### CODE REVIEW

File: docs/developer-guide/mapstore-migration-guide.md [hunk 1]
```diff
@@ -20,6 +20,17 @@ This is a list of things to check if you want to update from a previous version
...
+## Migration from 2023.02.xx to 2024.01.00
+
+### Annotations plugin refactor
+
+The Annotation plugin has been updated to be supported also in 3D maps. This update introduced some changes:
+
+- Configuration related to Annotation plugin has been removed from `localConfig.defaultState` and moved to the `cfg` property of the plugin
+- The annotations reducers is not needed anymore inside the default reducers of the app
+
+Please update the configuration if some customizations were applied to the Annotations plugin
```
Comment (offtherailz) [inline docs/developer-guide/mapstore-migration-guide.md:32]:
```suggestion
The Annotation plugin has been updated to be supported also in 3D maps. This update introduced some changes:

- All the configurations related to the "Annotations" plugin has been removed from `localConfig.json` `defaultState` entry and moved to the `cfg` property of the plugin
- The annotations reducers is not needed anymore inside the default reducers of the app

Please update the configuration if some customizations were applied to the Annotations plugin by: 
- Removing `annotations` entry from your `localConfig.json` `defaultState`
- If you customized the app, you can remove the `annotations` reducer from default reducers.
```

File: build/docma-config.json [hunk 2]
```diff
@@ -222,7 +222,7 @@
             "plugins": [
                 "web/client/plugins/index.jsdoc",
                 "web/client/plugins/AddGroup.jsx",
-                "web/client/plugins/Annotations.jsx",
+                "web/client/plugins/Annotations/index.jsx",
```
Comment (offtherailz) [inline build/docma-config.json:225]: Wrong file name caused the doc not to be added to the plugins documentation
```suggestion
                "web/client/plugins/Annotations/index.js",
```

File: web/client/plugins/Annotations/index.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  * symbols.json present in symbolsPath folder is mandatory and it contains the list of symbols to be used in the Annotations Plugin
+  * - width and height of SVGs should be 64px
+  * - the name is related to the filename of the s symbol
+  * - the label is used in the symbol dropdown menu
+  * [
+  *   {"name": "filename", "label": "label"},
+  *   {"name": "square", "label": "Square"}
+  * ]
+  *
+  * Example of plugin configuration
```
Comment (offtherailz) [inline web/client/plugins/Annotations/index.js:52]:
```suggestion
  * Annotations Plugin. Implements annotations handling on maps.
  * @prop {object[]} lineDashOptions `[{value: [line1 gap1 line2 gap2 line3...]}, {...}]` defines how dashed lines are displayed.
  * Use values without unit identifier.
  * If an odd number of values is inserted then they are added again to reach an even number of values
  * for more information see [this page](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/stroke-dasharray)
  * @prop {string} defaultPointType default point type of marker geometry type. Can be `'marker'` or `'symbol'`
  * @prop {string} defaultShape the default symbol used when switching for the symbol styler
  * @prop {string} defaultShapeStrokeColor default symbol stroke color
  * @prop {string} defaultShapeFillColor default symbol fill color
  * @prop {number} defaultShapeSize default symbol shape size in px
  * @prop {string} format `'decimal'` or `'aeronautical'` degree for coordinates
  * @prop {string} defaultTextAnnotation default text value for text annotations (default: `'New'`)
  * @prop {object} geometryEditorOptions properties to be passed to `CoordinatesEditor` of `GeometryEditor`. For more information refer to the documentation of `CoordinatesEditor` component
  * @prop {string} symbolsPath the relative path to the symbols folder where `symbols.json` and SVGs are located (starting from the `index.html` folder, i.e. the root) `symbols.json` can be structured like [this](https://github.com/geosolutions-it/MapStore2/blob/90fb33465fd3ff56c4bbaafb5ab0ed492826622c/web/client/product/assets/symbols/symbols.json)
  * `symbols.json` present in `symbolsPath` folder is mandatory and it contains the list of symbols to be used in the Annotations Plugin
  * - width and height of SVGs should be 64px
  * - the `name` is related to the filename of the symbol (the file must be named `<name>.svg`).
  * - the `label` is the name used in the symbol dropdown menu
  * ```
  * [
  *   {"name": "filename", "label": "label"},
  *   {"name": "square", "label": "Square"}
  * ]
  * ```
  * @prop {boolean} geodesic **deprecated** draw geodesic annotation. By default `geodesic` is `true` (Currently applicable only for "circle" annotation type)
  * @prop {object[]} fields **deprecated** Available annotation fields (`title` and `description` are mandatory). A list of object specifying:
  *  - `name`: the field synthetic name
  *  - `type`: type of value for the field (`'text'` or `'html'`)
  *  - `validator`: (optional) function for rule for validation
  *  - `validationError`: (optional) id for the translations file containing the validation error message
  *  - `showLabel`: whether to show or not the label of the field in the viewer / editor
  *  - `editable`: whether the field can be edited or not in editing mode
  * @class Annotations
  * @memberof plugins
  * @static
  * @example
  * // Example of plugin configuration
```

File: web/client/translations/data.it-IT.json [hunk 4]
```diff
@@ -1243,7 +1244,10 @@
             },
             "geometryError": "La configurazione della geometria non è valida",
             "annotationSaveGeometryError": "Il salvataggio è disabilitato a causa di una geometria non valida, correggilo per salvare l'annotazione",
-            "resolveAllErrors": "Si prega di correggere gli errori prima di modificare altre geometrie"
+            "resolveAllErrors": "Si prega di correggere gli errori prima di modificare altre geometrie",
+            "selectAnnotationFeature": "Seleziona una geometria annotation nel pannello di sinistra per editare coordinate e stili",
```
Comment (offtherailz) [inline web/client/translations/data.it-IT.json:1248]: If we use "si prega" just before, let's use the same level of confidence  for the other strings.
```suggestion
            "selectAnnotationFeature": "Selezionare una geometria annotation nel pannello di sinistra per editare coordinate e stili",
```

File: web/client/translations/data.it-IT.json [hunk 5]
```diff
@@ -1243,7 +1244,10 @@
...
+            "properties": "Proprietà",
+            "deactivatedMapInteraction": "Le interazioni di disegno delle annotazioni sono state disattivate perchè il '{pluginName}' plugin è stato attivato. Chiudi il '{pluginName}' plugin per attivare le interazioni di disegno delle annotazioni"
```
Comment (offtherailz) [inline web/client/translations/data.it-IT.json:1250]:
```suggestion
            "deactivatedMapInteraction": "Le interazioni di disegno delle annotazioni sono state disattivate perchè il '{pluginName}' plugin è stato attivato. Chiudere il '{pluginName}' plugin per attivare le interazioni di disegno delle annotazioni"
```

**Review by offtherailz — CHANGES_REQUESTED** [review]
First of all, I like the final result a lot. :+1: :+1: :+1: :rocket: 

Please organize with @tdipisa to check my notes and make some of them or open issues for things that can be postponed. 

- [ ] Inline fixes for documentation etc... 
- [ ] Identify on hover for annotations do not look to work anymore See this map: http://localhost:8081/#/viewer/36414

current branch: http://localhost:8081/#/viewer/36414

[identify.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/d19ef4aa-23d2-4b94-96bc-9f9a63671238)

dev: https://dev-mapstore.geosolutionsgroup.com/mapstore/#/viewer/36414

[identify_dev.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/21c4eb22-c563-4bc2-a573-f5d8826443bc)

Anyway I noticed, that after activating the identify on over, when switching to another map, the identify stops working at all, so we may need to create a separate issue for investigating issues with this tool.

Map exported 
[identify_map.zip](https://github.com/geosolutions-it/MapStore2/files/12584910/identify_map.zip)


- [ ] grab icon in TOC is not fully visible

image

- [ ] I was able to reproduce an invalid annotation here but I can not see the error details or how to fix it. (removing last coordinate fixed it).

[annotation_creation_error.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/9a1338f0-ee0d-4e3d-a2f9-57ca1653dc77)

image

- [ ] The default style for circle is a little unnatural ( 0px outline, gray fill) . It should be 1 px outline, to be consistent with the other styles.

- [ ] Height reference relative to the pointer hieight do not seems to work for Geometry transformation, if we do not set "bring to the front". I don't know, maybe it is a cesium issue.
[geometry_transformation.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/51ab7e30-20aa-45b2-a7e2-3e0675dd4bd6)

**Comment by tdipisa** [comment]
@offtherailz I've discussed points above with @allyoucanmap and he will check all of them today to see how to proceed with some of them (eg. the identify one). For sure we will create a separate issue for normalizing the grab icon.

**Comment by allyoucanmap** [comment]
@tdipisa @offtherailz opened a separated issue for the icon https://github.com/geosolutions-it/MapStore2/issues/9430

**Comment by allyoucanmap** [comment]
> - [ ]  Height reference relative to the pointer hieight do not seems to work for Geometry transformation, if we do not set "bring to the front". I don't know, maybe it is a cesium issue.

@offtherailz @tdipisa based on the video it seems that the marker is inside the building so it's not visible when bring to front is false. Bring to front property makes the marker visible even if behind another feature

**Comment by allyoucanmap** [comment]
> @offtherailz @tdipisa based on the video it seems that the marker is inside the building so it's not visible when bring to front is false. Bring to front property makes the marker visible even if behind another feature

After checking again it seems a problem related to the use of eyeOffset applied as workaround to manage the z-index of billboard, I'll open a new issue to investigate this

**Review by offtherailz — APPROVED** [review]
Tested, all looks good (and cool)

**Comment by allyoucanmap** [comment]
@ElenaGallo, please test this PR on dev, thanks

**Pull request merged**